compile.js

```diff
@@ -415,3 +462,72 @@ function getDefaultWriteTest(ctx, field) {
 function isPacked(field) {
     return field.options.packed === 'true';
 }
+
+function getTsType(field) {
+    let type = field.type;
+
+    if (fieldShouldUseStringAsNumber(field)) type = 'string';
+    else if (fieldTypeIsNumber(field)) type = 'number';
+    else if (field.type === 'bytes') type = 'Uint8Array';
+    else if (field.type === 'bool') type = 'boolean';
+
+    return field.repeated ? `Array<${type}>` : type;
+}
+
+function getMapTsType(fields) {
+    const key = getTsType(fields[0]);
+    const value = getTsType(fields[1]);
+    return {key, value};
+}
+
+/**
+ * @param {string} type
+ * @param {string} name
+ * @param {{name: string; type: string; required: boolean}} [fields]
+ * @returns {string}
+ */
+function typeDef(type, name, fields = []) {
+    const unionProperties = {};
+    const properties = fields.map((field) => {
+        if (field.oneof) {
+            unionProperties[field.oneof] = unionProperties[field.oneof] || [];
+            unionProperties[field.oneof].push(field.name);
+        }
+
+        const type = getTsType(field);
+        const isRequired = field.required || field.repeated || field.map;
+
+        const name = isRequired ? field.name : `[${field.name}]`;
+
+        return ` * @property {${type}} ${name}`;
+    });
+
+    for (const prop in unionProperties) {
+        const union = unionProperties[prop].map(s => `"${s}"`).join(' | ');
+        properties.push(` * @property {${union}} [${prop}]`);
+    }
+
+    return ['/**', ` * @typedef {${type}} ${name}`, ...properties, ' */']
+        .join('\n')
+        .concat('\n');
+}
+
+/**
+ * @param {object} ctx
+ * @param {string} name
+ * @param {{name: string; type: string; required: boolean}} [fields]
+ * @returns {string}
+ */
+function compileType(ctx, name, fields = []) {
+    if (ctx._proto.map) {
+        const {key, value} = getMapTsType(fields);
+        return typeDef(`Object<${key}, ${value}>`, name, []);
```

Comment:
Nit: let's use the same casing as in other places for consistency.

```suggestion
        return typeDef(`object<${key}, ${value}>`, name, []);
```

Reply:
Good catch! We couldn't use the `Object` type here since it's not a generic type.

At the same time, there is a subtle difference: the `object` type is more specific than `Object`. It represents any non-primitive type, excluding `null` and `undefined`, so I'm using it as a default type for all top-level types.

The type for the proto file dictionaries should be a generic type that could also be `Indexed`. We could use either `Record<K, V>` or `{[_: K]: V}` here. I'll use `Record` here, but it's a matter of taste, and I could switch to the `{}` notation if preferred.

Reply:
Right, `Record` sounds good to me.